### DISCUSSION TITLE
Abstracting rule precedence

### DISCUSSION THREAD

**Discussion by mikeaustin** [peggyjs/peggy#181] [Show and tell]
I've been using PEG.js on and off for years now, and I like the syntax and stability. One thing though is that it doesn't support left recursion, so all precedence-dependent rules are chained, which makes it difficult to re-arrange or remove rules. I recently experimented with simple precedence abstraction, so that rules can be defined independently of their precedence:

https://github.com/mikeaustin/kopi/blob/master/kopi4/src/rules.js

A trimmed down example below. I'm glad to see PEG.js revived. I feels it's the simplest and most straightforward PEG parser for JS.

```
const nonTerminals = {
  MultiplyExpression: NextRule => `
    = head:${NextRule} tail:(_ ("*" / "/") _ ${NextRule})* {
        ...
      }
  `,
  PrimaryExpression: NextRule => `
    = Identifier
  `
};

const terminals = {
  Identifier: `
    = [a-zA-Z][a-zA-Z0-9]* _ { return text(); }
  `
};

const precedence = [
  'MultiplyExpression',
  'PrimaryExpression'
];

const orderedNonTerminals = precedence.reverse().reduce(([orderedNonTerminals, nextRule], rule) => (
  [[...orderedNonTerminals, rule + nonTerminals[rule](nextRule)], rule]
), [[], null])[0].reverse();

const orderedTerminals = Object.entries(terminals).map(([name, rule]) => name + rule);

console.log(orderedNonTerminals.join('\n') + '\n' + orderedTerminals.join('\n'));
```

**Comment by Mingun** [comment]
The thing that you implemented looks like a semi-generic solution for the [template rules](https://github.com/pegjs/pegjs/issues/45) and if and when their will be implemented in the peggy it should solve your problem.
